# Pandas Dataframe

<img src="https://frenzy86.s3.eu-west-2.amazonaws.com/IFAO/dataframe.png" width="1200">

In [5]:
import numpy as np
import pandas as pd

In [6]:
np.random.seed(667)
mydata = np.random.randint(0,101,(4,3))

In [7]:
mydata

array([[ 85,  15,  89],
       [100,   3,  90],
       [  2,  64,   0],
       [ 36,  99,  47]])

In [8]:
myindex = ['CA','NY','AZ','TX']

In [9]:
mycolumns = ['Jan','Feb','Mar']

In [10]:
#pd.DataFrame()

In [11]:
df = pd.DataFrame(data=mydata)
df

0   1   2
0   85  15  89
1  100   3  90
2    2  64   0
3   36  99  47

In [12]:
df = pd.DataFrame(index=myindex,data=mydata)
df

0   1   2
CA   85  15  89
NY  100   3  90
AZ    2  64   0
TX   36  99  47

In [13]:
df = pd.DataFrame(myindex,mycolumns,mydata)


ValueError: Index data must be 1-dimensional

In [16]:
df = pd.DataFrame(data=mydata,index=myindex,columns=mycolumns)
df

Jan  Feb  Mar
CA   85   15   89
NY  100    3   90
AZ    2   64    0
TX   36   99   47

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, CA to TX
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Jan     4 non-null      int64
 1   Feb     4 non-null      int64
 2   Mar     4 non-null      int64
dtypes: int64(3)
memory usage: 128.0+ bytes


In [18]:
df['Jan']

CA     85
NY    100
AZ      2
TX     36
Name: Jan, dtype: int64

In [19]:
df['Jan'].tolist()

[85, 100, 2, 36]

In [20]:
data = {'Nome':['Gianni','Enrico','Sergio','Andrea','Marco'],
        'Eta': [23,56,32,None,35],
        'Professione':['cuoco','postino','manager','sbirro','inoccupato']
        }

In [21]:
df1 = pd.DataFrame(data)
df1

Nome   Eta Professione
0  Gianni  23.0       cuoco
1  Enrico  56.0     postino
2  Sergio  32.0     manager
3  Andrea   NaN      sbirro
4   Marco  35.0  inoccupato

In [22]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Nome         5 non-null      object 
 1   Eta          4 non-null      float64
 2   Professione  5 non-null      object 
dtypes: float64(1), object(2)
memory usage: 248.0+ bytes


## Read external data tables

In [23]:
!wget https://frenzy86.s3.eu-west-2.amazonaws.com/IFAO/tips_complete.csv

--2024-01-26 09:05:04--  https://frenzy86.s3.eu-west-2.amazonaws.com/IFAO/tips_complete.csv
Resolving frenzy86.s3.eu-west-2.amazonaws.com (frenzy86.s3.eu-west-2.amazonaws.com)... 52.95.149.178, 3.5.246.197, 52.95.150.146, ...
Connecting to frenzy86.s3.eu-west-2.amazonaws.com (frenzy86.s3.eu-west-2.amazonaws.com)|52.95.149.178|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18752 (18K) [text/csv]
Saving to: ‘tips_complete.csv’

tips_complete.csv   100%[===================>]  18.31K  --.-KB/s    in 0.001s  

2024-01-26 09:05:05 (26.4 MB/s) - ‘tips_complete.csv’ saved [18752/18752]



In [24]:
ls

sample_data/  tips_complete.csv


In [25]:
df = pd.read_csv('tips_complete.csv')
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

----
About this DataSet (in case you are interested)

* Description
    * One waiter recorded information about each tip he received over a period of a few months working in one restaurant. He collected several variables:

* Format
    * A data frame with 244 rows and 7 variables

* Details
    * tip in dollars,
    * bill in dollars,
    * sex of the bill payer,
    * whether there were smokers in the party,
    * day of the week,
    * time of day,
    * size of the party.

In all he recorded 244 tips.
    
* Note: I created some additional columns with Fake data, including Name, CC Number, and Payment ID.

----

# DataFrames

In [26]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

In [27]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [28]:
df.head(10)

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   
5       25.29  4.71    Male     No  Sun  Dinner     4              6.32   
6        8.77  2.00    Male     No  Sun  Dinner     2              4.38   
7       26.88  3.12    Male     No  Sun  Dinner     4              6.72   
8       15.04  1.96    Male     No  Sun  Dinner     2              7.52   
9       14.78  3.23    Male     No  Sun  Dinner     2              7.39   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251  
5          Erik Smith   213140353657882    Sun9679  
6  Kristopher Johnson  2223727524230344    Sun5985  
7         Robert Buck  3514785077705092    Sun8157  
8     Joseph Mcdonald  3522866365840377    Sun6820  
9       Jerome Abbott  3532124519049786    Sun3775

In [29]:
df.tail(3)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

          Payer Name         CC Number Payment ID  
241       Keith Wong  6011891618747196    Sat3880  
242     Dennis Dixon     4375220550950      Sat17  
243  Michelle Hardin  3511451626698139    Thur672

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    int64  
 10  Payment ID        244 non-null    object 
dtypes: float64(3), int64(2), object(6)
memory usage: 21.1+ KB


In [31]:
len(df)#solo numero di righe

244

In [32]:
df.shape

(244, 11)

In [33]:
df.describe().T

count          mean           std           min  \
total_bill        244.0  1.978594e+01  8.902412e+00  3.070000e+00   
tip               244.0  2.998279e+00  1.383638e+00  1.000000e+00   
size              244.0  2.569672e+00  9.510998e-01  1.000000e+00   
price_per_person  244.0  7.888197e+00  2.914234e+00  2.880000e+00   
CC Number         244.0  2.563496e+15  2.369340e+15  6.040679e+10   

                           25%           50%           75%           max  
total_bill        1.334750e+01  1.779500e+01  2.412750e+01  5.081000e+01  
tip               2.000000e+00  2.900000e+00  3.562500e+00  1.000000e+01  
size              2.000000e+00  2.000000e+00  3.000000e+00  6.000000e+00  
price_per_person  5.800000e+00  7.255000e+00  9.390000e+00  2.027000e+01  
CC Number         3.040731e+13  3.525318e+15  4.553675e+15  6.596454e+15

In [34]:
#df.describe().transpose()
df.describe().T

count          mean           std           min  \
total_bill        244.0  1.978594e+01  8.902412e+00  3.070000e+00   
tip               244.0  2.998279e+00  1.383638e+00  1.000000e+00   
size              244.0  2.569672e+00  9.510998e-01  1.000000e+00   
price_per_person  244.0  7.888197e+00  2.914234e+00  2.880000e+00   
CC Number         244.0  2.563496e+15  2.369340e+15  6.040679e+10   

                           25%           50%           75%           max  
total_bill        1.334750e+01  1.779500e+01  2.412750e+01  5.081000e+01  
tip               2.000000e+00  2.900000e+00  3.562500e+00  1.000000e+01  
size              2.000000e+00  2.000000e+00  3.000000e+00  6.000000e+00  
price_per_person  5.800000e+00  7.255000e+00  9.390000e+00  2.027000e+01  
CC Number         3.040731e+13  3.525318e+15  4.553675e+15  6.596454e+15

In [35]:
df.sum()

total_bill                                                    4827.77
tip                                                            731.58
sex                 FemaleMaleMaleMaleFemaleMaleMaleMaleMaleMaleMa...
smoker              NoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNo...
day                 SunSunSunSunSunSunSunSunSunSunSunSunSunSunSunS...
time                DinnerDinnerDinnerDinnerDinnerDinnerDinnerDinn...
size                                                              627
price_per_person                                              1924.72
Payer Name          Christy CunninghamDouglas TuckerTravis Walters...
CC Number                                          625492917132785482
Payment ID          Sun2959Sun4608Sun4458Sun5260Sun2251Sun9679Sun5...
dtype: object

## Selection and Indexing

Let's learn how to retrieve information from a DataFrame.

### COLUMNS

We will begin be learning how to extract information based on the columns

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

#### Grab a Single Column

In [ ]:
df['total_bill']

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

In [ ]:
type(df['total_bill'])

pandas.core.series.Series

In [ ]:
df[['tip','total_bill']].mean()

tip            2.998279
total_bill    19.785943
dtype: float64

In [ ]:
df.total_bill

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

#### Grab Multiple Columns

In [ ]:
# Note how its a python list of column names! Thus the double brackets.
df[['total_bill','tip']] #è una lista di selezione!!!

total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

#### Create New Columns

In [ ]:
df['new'] = 2
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  cost  new  
0    Christy Cunningham  3560325168603410    Sun2959     2    2  
1        Douglas Tucker  4478071379779230    Sun4608     2    2  
2        Travis Walters  6011812112971322    Sun4458     2    2  
3      Nathaniel Harris  4676137647685994    Sun5260     2    2  
4          Tonya Carter  4832732618637221    Sun2251     2    2  
..                  ...               ...        ...   ...  ...  
239       Michael Avila  5296068606052842    Sat2657     2    2  
240      Monica Sanders  3506806155565404    Sat1766     2    2  
241          Keith Wong  6011891618747196    Sat3880     2    2  
242        Dennis Dixon     4375220550950      Sat17     2    2  
243     Michelle Hardin  3511451626698139    Thur672     2    2  

[244 rows x 13 columns]

In [ ]:
100* df['tip'] / df['total_bill']

0       5.944673
1      16.054159
2      16.658734
3      13.978041
4      14.680765
         ...    
239    20.392697
240     7.358352
241     8.822232
242     9.820426
243    15.974441
Length: 244, dtype: float64

In [ ]:
df['tip_percentage'] = 100* df['tip'] / df['total_bill']

In [ ]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  tip_percentage  
0    Christy Cunningham  3560325168603410    Sun2959        5.944673  
1        Douglas Tucker  4478071379779230    Sun4608       16.054159  
2        Travis Walters  6011812112971322    Sun4458       16.658734  
3      Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4          Tonya Carter  4832732618637221    Sun2251       14.680765  
..                  ...               ...        ...             ...  
239       Michael Avila  5296068606052842    Sat2657       20.392697  
240      Monica Sanders  3506806155565404    Sat1766        7.358352  
241          Keith Wong  6011891618747196    Sat3880        8.822232  
242        Dennis Dixon     4375220550950      Sat17        9.820426  
243     Michelle Hardin  3511451626698139    Thur672       15.974441  

[244 rows x 12 columns]

In [ ]:
df['price_per_person'] = df['total_bill'] / df['size']

In [ ]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2          8.495000   
1         10.34  1.66    Male     No   Sun  Dinner     3          3.446667   
2         21.01  3.50    Male     No   Sun  Dinner     3          7.003333   
3         23.68  3.31    Male     No   Sun  Dinner     2         11.840000   
4         24.59  3.61  Female     No   Sun  Dinner     4          6.147500   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3          9.676667   
240       27.18  2.00  Female    Yes   Sat  Dinner     2         13.590000   
241       22.67  2.00    Male    Yes   Sat  Dinner     2         11.335000   
242       17.82  1.75    Male     No   Sat  Dinner     2          8.910000   
243       18.78  3.00  Female     No  Thur  Dinner     2          9.390000   

             Payer Name         CC Number Payment ID  cost  new  
0    Christy Cunningham  3560325168603410    Sun2959     2    2  
1        Douglas Tucker  4478071379779230    Sun4608     2    2  
2        Travis Walters  6011812112971322    Sun4458     2    2  
3      Nathaniel Harris  4676137647685994    Sun5260     2    2  
4          Tonya Carter  4832732618637221    Sun2251     2    2  
..                  ...               ...        ...   ...  ...  
239       Michael Avila  5296068606052842    Sat2657     2    2  
240      Monica Sanders  3506806155565404    Sat1766     2    2  
241          Keith Wong  6011891618747196    Sat3880     2    2  
242        Dennis Dixon     4375220550950      Sat17     2    2  
243     Michelle Hardin  3511451626698139    Thur672     2    2  

[244 rows x 13 columns]

In [ ]:
help(np.round)

Help on function round_ in module numpy:

round_(a, decimals=0, out=None)
    Round an array to the given number of decimals.
    
    See Also
    --------
    around : equivalent function; see for details.



#### Adjust Existing Columns

In [ ]:
# Because pandas is based on numpy, we get awesome capabilities with numpy's universal functions!
df['price_per_person'] = np.round(df['price_per_person'],2)

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

#### Remove Columns

In [ ]:
df.drop('tip_percentage',axis=1)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  cost  new  
0    Christy Cunningham  3560325168603410    Sun2959     2    2  
1        Douglas Tucker  4478071379779230    Sun4608     2    2  
2        Travis Walters  6011812112971322    Sun4458     2    2  
3      Nathaniel Harris  4676137647685994    Sun5260     2    2  
4          Tonya Carter  4832732618637221    Sun2251     2    2  
..                  ...               ...        ...   ...  ...  
239       Michael Avila  5296068606052842    Sat2657     2    2  
240      Monica Sanders  3506806155565404    Sat1766     2    2  
241          Keith Wong  6011891618747196    Sat3880     2    2  
242        Dennis Dixon     4375220550950      Sat17     2    2  
243     Michelle Hardin  3511451626698139    Thur672     2    2  

[244 rows x 13 columns]

In [ ]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  tip_percentage  
0    Christy Cunningham  3560325168603410    Sun2959        5.944673  
1        Douglas Tucker  4478071379779230    Sun4608       16.054159  
2        Travis Walters  6011812112971322    Sun4458       16.658734  
3      Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4          Tonya Carter  4832732618637221    Sun2251       14.680765  
..                  ...               ...        ...             ...  
239       Michael Avila  5296068606052842    Sat2657       20.392697  
240      Monica Sanders  3506806155565404    Sat1766        7.358352  
241          Keith Wong  6011891618747196    Sat3880        8.822232  
242        Dennis Dixon     4375220550950      Sat17        9.820426  
243     Michelle Hardin  3511451626698139    Thur672       15.974441  

[244 rows x 12 columns]

In [ ]:
df = df.drop("tip_percentage",axis=1) #sovrascrivere!!!!!!!

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

# Index Basics

Before going over the same retrieval tasks for rows, let's build some basic understanding of the pandas DataFrame Index.

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [ ]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [ ]:
df.set_index('Payment ID')

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
...                ...   ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  \
Payment ID                                                           
Sun2959                 8.49  Christy Cunningham  3560325168603410   
Sun4608                 3.45      Douglas Tucker  4478071379779230   
Sun4458                 7.00      Travis Walters  6011812112971322   
Sun5260                11.84    Nathaniel Harris  4676137647685994   
Sun2251                 6.15        Tonya Carter  4832732618637221   
...                      ...                 ...               ...   
Sat2657                 9.68       Michael Avila  5296068606052842   
Sat1766                13.59      Monica Sanders  3506806155565404   
Sat3880                11.34          Keith Wong  6011891618747196   
Sat17                   8.91        Dennis Dixon     4375220550950   
Thur672                 9.39     Michelle Hardin  3511451626698139   

            tip_percentage  
Payment ID                  
Sun2959           5.944673  
Sun4608          16.054159  
Sun4458          16.658734  
Sun5260          13.978041  
Sun2251          14.680765  
...                    ...  
Sat2657          20.392697  
Sat1766           7.358352  
Sat3880           8.822232  
Sat17             9.820426  
Thur672          15.974441  

[244 rows x 11 columns]

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

In [ ]:
df = df.set_index('Payment ID')#sovrascrivere!!!!!!!
df

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
...                ...   ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  cost  new  \
Payment ID                                                                      
Sun2959                 8.49  Christy Cunningham  3560325168603410     2    2   
Sun4608                 3.45      Douglas Tucker  4478071379779230     2    2   
Sun4458                 7.00      Travis Walters  6011812112971322     2    2   
Sun5260                11.84    Nathaniel Harris  4676137647685994     2    2   
Sun2251                 6.15        Tonya Carter  4832732618637221     2    2   
...                      ...                 ...               ...   ...  ...   
Sat2657                 9.68       Michael Avila  5296068606052842     2    2   
Sat1766                13.59      Monica Sanders  3506806155565404     2    2   
Sat3880                11.34          Keith Wong  6011891618747196     2    2   
Sat17                   8.91        Dennis Dixon     4375220550950     2    2   
Thur672                 9.39     Michelle Hardin  3511451626698139     2    2   

            tip_percentage  
Payment ID                  
Sun2959           5.944673  
Sun4608          16.054159  
Sun4458          16.658734  
Sun5260          13.978041  
Sun2251          14.680765  
...                    ...  
Sat2657          20.392697  
Sat1766           7.358352  
Sat3880           8.822232  
Sat17             9.820426  
Thur672          15.974441  

[244 rows x 13 columns]

In [ ]:
df = df.set_index('tip')#sovrascrivere!!!!!!!
df

Payment ID  total_bill     sex smoker   day    time  size  \
tip                                                              
1.01    Sun2959       16.99  Female     No   Sun  Dinner     2   
1.66    Sun4608       10.34    Male     No   Sun  Dinner     3   
3.50    Sun4458       21.01    Male     No   Sun  Dinner     3   
3.31    Sun5260       23.68    Male     No   Sun  Dinner     2   
3.61    Sun2251       24.59  Female     No   Sun  Dinner     4   
...         ...         ...     ...    ...   ...     ...   ...   
5.92    Sat2657       29.03    Male     No   Sat  Dinner     3   
2.00    Sat1766       27.18  Female    Yes   Sat  Dinner     2   
2.00    Sat3880       22.67    Male    Yes   Sat  Dinner     2   
1.75      Sat17       17.82    Male     No   Sat  Dinner     2   
3.00    Thur672       18.78  Female     No  Thur  Dinner     2   

      price_per_person          Payer Name         CC Number  cost  new  \
tip                                                                       
1.01              8.49  Christy Cunningham  3560325168603410     2    2   
1.66              3.45      Douglas Tucker  4478071379779230     2    2   
3.50              7.00      Travis Walters  6011812112971322     2    2   
3.31             11.84    Nathaniel Harris  4676137647685994     2    2   
3.61              6.15        Tonya Carter  4832732618637221     2    2   
...                ...                 ...               ...   ...  ...   
5.92              9.68       Michael Avila  5296068606052842     2    2   
2.00             13.59      Monica Sanders  3506806155565404     2    2   
2.00             11.34          Keith Wong  6011891618747196     2    2   
1.75              8.91        Dennis Dixon     4375220550950     2    2   
3.00              9.39     Michelle Hardin  3511451626698139     2    2   

      tip_percentage  
tip                   
1.01        5.944673  
1.66       16.054159  
3.50       16.658734  
3.31       13.978041  
3.61       14.680765  
...              ...  
5.92       20.392697  
2.00        7.358352  
2.00        8.822232  
1.75        9.820426  
3.00       15.974441  

[244 rows x 13 columns]

In [ ]:
df = df.reset_index()
df

tip Payment ID  total_bill     sex smoker   day    time  size  \
0    1.01    Sun2959       16.99  Female     No   Sun  Dinner     2   
1    1.66    Sun4608       10.34    Male     No   Sun  Dinner     3   
2    3.50    Sun4458       21.01    Male     No   Sun  Dinner     3   
3    3.31    Sun5260       23.68    Male     No   Sun  Dinner     2   
4    3.61    Sun2251       24.59  Female     No   Sun  Dinner     4   
..    ...        ...         ...     ...    ...   ...     ...   ...   
239  5.92    Sat2657       29.03    Male     No   Sat  Dinner     3   
240  2.00    Sat1766       27.18  Female    Yes   Sat  Dinner     2   
241  2.00    Sat3880       22.67    Male    Yes   Sat  Dinner     2   
242  1.75      Sat17       17.82    Male     No   Sat  Dinner     2   
243  3.00    Thur672       18.78  Female     No  Thur  Dinner     2   

     price_per_person          Payer Name         CC Number  cost  new  \
0                8.49  Christy Cunningham  3560325168603410     2    2   
1                3.45      Douglas Tucker  4478071379779230     2    2   
2                7.00      Travis Walters  6011812112971322     2    2   
3               11.84    Nathaniel Harris  4676137647685994     2    2   
4                6.15        Tonya Carter  4832732618637221     2    2   
..                ...                 ...               ...   ...  ...   
239              9.68       Michael Avila  5296068606052842     2    2   
240             13.59      Monica Sanders  3506806155565404     2    2   
241             11.34          Keith Wong  6011891618747196     2    2   
242              8.91        Dennis Dixon     4375220550950     2    2   
243              9.39     Michelle Hardin  3511451626698139     2    2   

     tip_percentage  
0          5.944673  
1         16.054159  
2         16.658734  
3         13.978041  
4         14.680765  
..              ...  
239       20.392697  
240        7.358352  
241        8.822232  
242        9.820426  
243       15.974441  

[244 rows x 14 columns]

### ROWS

Let's now explore these same concepts but with Rows.

#### Grab a Single Row

In [ ]:
# Integer Based
df.iloc[0] # prende la riga con indice indicato

tip                               1.01
Payment ID                     Sun2959
total_bill                       16.99
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
cost                                 2
new                                  2
tip_percentage                5.944673
Name: 0, dtype: object

In [ ]:
# Name Based
df = df.set_index('Payment ID')
df.loc['Sun4608'] # pesca dall'index

tip                             1.66
total_bill                     10.34
sex                             Male
smoker                            No
day                              Sun
time                          Dinner
size                               3
price_per_person                3.45
Payer Name            Douglas Tucker
CC Number           4478071379779230
cost                               2
new                                2
tip_percentage             16.054159
Name: Sun4608, dtype: object

#### Grab Multiple Rows

In [ ]:
df.iloc[0:4] #prime 4 righe

tip  total_bill     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959     1.01       16.99  Female     No  Sun  Dinner     2   
Sun4608     1.66       10.34    Male     No  Sun  Dinner     3   
Sun4458     3.50       21.01    Male     No  Sun  Dinner     3   
Sun5260     3.31       23.68    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  cost  new  \
Payment ID                                                                      
Sun2959                 8.49  Christy Cunningham  3560325168603410     2    2   
Sun4608                 3.45      Douglas Tucker  4478071379779230     2    2   
Sun4458                 7.00      Travis Walters  6011812112971322     2    2   
Sun5260                11.84    Nathaniel Harris  4676137647685994     2    2   

            tip_percentage  
Payment ID                  
Sun2959           5.944673  
Sun4608          16.054159  
Sun4458          16.658734  
Sun5260          13.978041

In [ ]:
df.loc[['Sun2959','Sun5260']]

tip  total_bill     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959     1.01       16.99  Female     No  Sun  Dinner     2   
Sun5260     3.31       23.68    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  cost  new  \
Payment ID                                                                      
Sun2959                 8.49  Christy Cunningham  3560325168603410     2    2   
Sun5260                11.84    Nathaniel Harris  4676137647685994     2    2   

            tip_percentage  
Payment ID                  
Sun2959           5.944673  
Sun5260          13.978041

In [ ]:
df.iloc[:,-2:]

new  tip_percentage
Payment ID                     
Sun2959       2        5.944673
Sun4608       2       16.054159
Sun4458       2       16.658734
Sun5260       2       13.978041
Sun2251       2       14.680765
...         ...             ...
Sat2657       2       20.392697
Sat1766       2        7.358352
Sat3880       2        8.822232
Sat17         2        9.820426
Thur672       2       15.974441

[244 rows x 2 columns]

#### Remove Row

Typically are datasets will be large enough that we won't remove rows like this since we won't know thier row location for some specific condition, instead, we drop rows based on conditions such as missing data or column values. The next lecture will cover this in a lot more detail.

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  \
Payment ID                                                           
Sun2959                 8.49  Christy Cunningham  3560325168603410   
Sun4608                 3.45      Douglas Tucker  4478071379779230   
Sun4458                 7.00      Travis Walters  6011812112971322   
Sun5260                11.84    Nathaniel Harris  4676137647685994   
Sun2251                 6.15        Tonya Carter  4832732618637221   

            tip_percentage  
Payment ID                  
Sun2959           5.944673  
Sun4608          16.054159  
Sun4458          16.658734  
Sun5260          13.978041  
Sun2251          14.680765

In [ ]:
df.drop('Sun2959',axis=0).head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   
Sun9679          25.29  4.71    Male     No  Sun  Dinner     4   

            price_per_person        Payer Name         CC Number  \
Payment ID                                                         
Sun4608                 3.45    Douglas Tucker  4478071379779230   
Sun4458                 7.00    Travis Walters  6011812112971322   
Sun5260                11.84  Nathaniel Harris  4676137647685994   
Sun2251                 6.15      Tonya Carter  4832732618637221   
Sun9679                 6.32        Erik Smith   213140353657882   

            tip_percentage  
Payment ID                  
Sun4608          16.054159  
Sun4458          16.658734  
Sun5260          13.978041  
Sun2251          14.680765  
Sun9679          18.623962

In [ ]:
# Error if you have a named index!
# df.drop(0,axis=0).head()

#### Insert a New Row

Pretty rare to add a single row like this. Usually you use pd.concat() to add many rows at once. You could use the .append() method with a list of pd.Series() objects, but you won't see us do this with realistic real-world data.

In [ ]:
one_row = df.iloc[0]

In [ ]:
one_row

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
tip_percentage                5.944673
Name: Sun2959, dtype: object

In [ ]:
type(one_row)

pandas.core.series.Series

In [ ]:
df.tail()

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person       Payer Name         CC Number  \
Payment ID                                                        
Sat2657                 9.68    Michael Avila  5296068606052842   
Sat1766                13.59   Monica Sanders  3506806155565404   
Sat3880                11.34       Keith Wong  6011891618747196   
Sat17                   8.91     Dennis Dixon     4375220550950   
Thur672                 9.39  Michelle Hardin  3511451626698139   

            tip_percentage  
Payment ID                  
Sat2657          20.392697  
Sat1766           7.358352  
Sat3880           8.822232  
Sat17             9.820426  
Thur672          15.974441

In [ ]:
df.append(one_row).tail()

<ipython-input-72-161d04e87f2b>:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df.append(one_row).tail()


tip  total_bill     sex smoker   day    time  size  \
Payment ID                                                        
Sat1766     2.00       27.18  Female    Yes   Sat  Dinner     2   
Sat3880     2.00       22.67    Male    Yes   Sat  Dinner     2   
Sat17       1.75       17.82    Male     No   Sat  Dinner     2   
Thur672     3.00       18.78  Female     No  Thur  Dinner     2   
Sun2959     1.01       16.99  Female     No   Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  cost  new  \
Payment ID                                                                      
Sat1766                13.59      Monica Sanders  3506806155565404     2    2   
Sat3880                11.34          Keith Wong  6011891618747196     2    2   
Sat17                   8.91        Dennis Dixon     4375220550950     2    2   
Thur672                 9.39     Michelle Hardin  3511451626698139     2    2   
Sun2959                 8.49  Christy Cunningham  3560325168603410     2    2   

            tip_percentage  
Payment ID                  
Sat1766           7.358352  
Sat3880           8.822232  
Sat17             9.820426  
Thur672          15.974441  
Sun2959           5.944673

--------

#### Extract an array

In [ ]:
array_row = df.iloc[0].values
array_row

array([16.99, 1.01, 'Female', 'No', 'Sun', 'Dinner', 2, 8.49,
       'Christy Cunningham', 3560325168603410], dtype=object)

In [ ]:
df['total_bill'].tolist()

[16.99,
 10.34,
 21.01,
 23.68,
 24.59,
 25.29,
 8.77,
 26.88,
 15.04,
 14.78,
 10.27,
 35.26,
 15.42,
 18.43,
 14.83,
 21.58,
 10.33,
 16.29,
 16.97,
 20.65,
 17.92,
 20.29,
 15.77,
 39.42,
 19.82,
 17.81,
 13.37,
 12.69,
 21.7,
 19.65,
 9.55,
 18.35,
 15.06,
 20.69,
 17.78,
 24.06,
 16.31,
 16.93,
 18.69,
 31.27,
 16.04,
 17.46,
 13.94,
 9.68,
 30.4,
 18.29,
 22.23,
 32.4,
 28.55,
 18.04,
 12.54,
 10.29,
 34.81,
 9.94,
 25.56,
 19.49,
 38.01,
 26.41,
 11.24,
 48.27,
 20.29,
 13.81,
 11.02,
 18.29,
 17.59,
 20.08,
 16.45,
 3.07,
 20.23,
 15.01,
 12.02,
 17.07,
 26.86,
 25.28,
 14.73,
 10.51,
 17.92,
 27.2,
 22.76,
 17.29,
 19.44,
 16.66,
 10.07,
 32.68,
 15.98,
 34.83,
 13.03,
 18.28,
 24.71,
 21.16,
 28.97,
 22.49,
 5.75,
 16.32,
 22.75,
 40.17,
 27.28,
 12.03,
 21.01,
 12.46,
 11.35,
 15.38,
 44.3,
 22.42,
 20.92,
 15.36,
 20.49,
 25.21,
 18.24,
 14.31,
 14.0,
 7.25,
 38.07,
 23.95,
 25.71,
 17.31,
 29.93,
 10.65,
 12.43,
 24.08,
 11.69,
 13.42,
 14.26,
 15.95,
 12.48,
 29.8,
 8.52,